pyomo/core/plugins/transform/scaling.py

```diff
@@ -0,0 +1,255 @@
+        >>> print(value(model.scaled_x))
+        1.0
+        >>> print(value(scaled_model.scaled_x))
+        0.2
+        >>> print(value(scaled_model.scaled_x.lb))
+        -1.0
+        >>> print(value(model.scaled_obj))
+        101000000.0
+        >>> print(value(scaled_model.scaled_obj))
+        101.0
+
+    ToDo
+    ====
+    - implement an option to change the variables names or not
+
+    """
+
+    def __init__(self, **kwds):
+        kwds['name'] = "scale_model"
+        self._scaling_method = kwds.pop('scaling_method', 'user')
+        super(ScaleModel, self).__init__(**kwds)
+
+    def _create_using(self, original_model, **kwds):
+        scaled_model = original_model.clone()
+        self._apply_to(scaled_model, **kwds)
+        return scaled_model
+
+    def _get_float_scaling_factor(self, instance, component_data):
+        scaling_factor = None
+        if component_data in instance.scaling_factor:
+            scaling_factor = instance.scaling_factor[component_data]
+        elif component_data.parent_component() in instance.scaling_factor:
+            scaling_factor = instance.scaling_factor[component_data.parent_component()]
+
+        if scaling_factor is None:
+            return 1.0
+
+        try:
+            scaling_factor = float(scaling_factor)
+        except ValueError:
+            raise ValueError(
+                "Suffix 'scaling_factor' has a value %s for component %s that cannot be converted to a float. "
+                "Floating point values are required for this suffix in the ScaleModel transformation."
+                % (scaling_factor, component_data))
+        return scaling_factor
+
+    def _apply_to(self, model, **kwds):
+        # create a map of component to scaling factor
+        component_scaling_factor_map = ComponentMap()
+
+        # if the scaling_method is 'user', get the scaling parameters from the suffixes
+        if self._scaling_method == 'user':
+            # perform some checks to make sure we have the necessary suffixes
+            if type(model.component('scaling_factor')) is not Suffix:
+                raise ValueError("ScaleModel transformation called with scaling_method='user'"
+                                 ", but cannot find the suffix 'scaling_factor' on the model")
+
+            # get the scaling factors
+            for c in model.component_data_objects(ctype=(Var, Constraint, Objective), descend_into=True):
+                component_scaling_factor_map[c] = self._get_float_scaling_factor(model, c)
+        else:
+            raise ValueError("ScaleModel transformation: unknown scaling_method found"
+                             "-- supported values: 'user' ")
+
+        # rename all the Vars, Constraints, and Objectives from foo to scaled_foo
+        scaled_component_to_original_name_map = \
+            rename_components(model=model,
+                              component_list=list(model.component_objects(ctype=[Var, Constraint, Objective])),
```

Comment:
Why rename all Var, Constraint, and Objective components instead of only the ones found in the `scaling_factor` suffix?

Reply:
Just for consistency. I didn't want to keep a map of which were scaled and which were not. I assume if the user has specified scaling parameters for some variables and not others that the others are scaled to 1.0.